pkg/registry/service/allocator/controller/repair.go

```diff
@@ -0,0 +1,111 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/service"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/service/allocator"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+)
+
+// Repair is a controller loop that periodically examines all service PortalIP allocations
+// and logs any errors.
+// TODO: allocate new IPs if necessary
+// TODO: perform repair?
+type Repair struct {
+	interval time.Duration
+	registry service.Registry
+	network  *net.IPNet
+	alloc    service.IPRegistry
+}
+
+// NewRepair creates a controller that periodically ensures that all portalIPs are uniquely allocated across the cluster
+// and generates informational warnings for a cluster that is not in sync.
+func NewRepair(interval time.Duration, registry service.Registry, network *net.IPNet, alloc service.IPRegistry) *Repair {
+	return &Repair{
+		interval: interval,
+		registry: registry,
+		network:  network,
+		alloc:    alloc,
+	}
+}
+
+// RunUntil starts the controller until the provided ch is closed.
+func (c *Repair) RunUntil(ch chan struct{}) {
+	util.Until(func() {
+		if err := c.RunOnce(); err != nil {
+			util.HandleError(err)
+		}
+	}, c.interval, ch)
+}
+
+// RunOnce verifies the state of the portal IP allocations and returns an error if an unrecoverable problem occurs.
+func (c *Repair) RunOnce() error {
+	latest, err := c.alloc.Get()
+	if err != nil {
+		return fmt.Errorf("unable to refresh the service IP block: %v", err)
+	}
+
+	ctx := api.WithNamespace(api.NewDefaultContext(), api.NamespaceAll)
+	list, err := c.registry.ListServices(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to refresh the service IP block: %v", err)
+	}
+
+	r := allocator.NewCIDRRange(c.network)
+	for _, svc := range list.Items {
+		if !api.IsServiceIPSet(&svc) {
+			continue
+		}
+		ip := net.ParseIP(svc.Spec.PortalIP)
+		if ip == nil {
+			// portal IP is broken, reallocate
+			util.HandleError(fmt.Errorf("the portal IP %s for service %s/%s is not a valid IP; please recreate", svc.Spec.PortalIP, svc.Name, svc.Namespace))
+			continue
+		}
+		switch err := r.Allocate(ip); err {
+		case nil:
+		case allocator.ErrAllocated:
+			// TODO: send event
+			// portal IP is broken, reallocate
+			util.HandleError(fmt.Errorf("the portal IP %s for service %s/%s was assigned to multiple services; please recreate", ip, svc.Name, svc.Namespace))
```

Comment:
Right now it would be legacy data - when this change happens we will block creations of new services until the first repair reads the entire space and writes the allocations, and after that it should be fine.  

It's also a way to detect inaccuracies - we could turn this down to once an hour or once a day and it's really just a sanity check at that point.

> On May 4, 2015, at 8:07 PM, Alex Robinson <notifications@github.com> wrote:
> 
> In pkg/registry/service/allocator/controller/repair.go:
> 
> > +	for _, svc := range list.Items {
> > +		if !api.IsServiceIPSet(&svc) {
> > +			continue
> > +		}
> > +		ip := net.ParseIP(svc.Spec.PortalIP)
> > +		if ip == nil {
> > +			// portal IP is broken, reallocate
> > +			util.HandleError(fmt.Errorf("the portal IP %s for service %s/%s is not a valid IP; please recreate", svc.Spec.PortalIP, svc.Name, svc.Namespace))
> > +			continue
> > +		}
> > +		switch err := r.Allocate(ip); err {
> > +		case nil:
> > +		case allocator.ErrAllocated:
> > +			// TODO: send event
> > +			// portal IP is broken, reallocate
> > +			util.HandleError(fmt.Errorf("the portal IP %s for service %s/%s was assigned to multiple services; please recreate", ip, svc.Name, svc.Namespace))
> Under what circumstances can this occur? I haven't made it to all the code yet, but I thought you were doing a compare-and-swap against etcd, and so should be able to avoid conflicts like this.
> 
> —
> Reply to this email directly or view it on GitHub.
>